### PR TITLE
Support more human friendly k/d/x-line duration format

### PR DESCRIPTION
**Pull request by examknow**
Per #85

K/D/X-Line durations can now be set like `KLINE 3w2d1m *@1.2.3.4 :test` and have the same effect as `KLINE 33121 *@1.2.3.4 :test`. K-Lines set using the old duration format will still work the same as before.

### CODE REVIEW

File: ircd/s_newconf.c [hunk 1]
```diff
@@ -685,23 +685,39 @@ time_t
...
+	while (*ch) {
+		och = ch;
+		current = strtol(ch, &ch, 10);
+		if (och == ch)
+			/* No numbers were given so return invalid */
 			return -1;
+		if (current) {
+			switch (*ch++) {
+				case '\0':
+					/* No unit was given so send it back as seconds */
```
Comment (edk0) [inline ircd/s_newconf.c:700]: comment says seconds, we interpret it as minutes. confusing?

File: ircd/s_newconf.c [hunk 2]
```diff
@@ -685,23 +685,39 @@ time_t
...
+					break;
+				default:
+					/* Bad time unit was given */
+					return -1;
+			}
+		}
 	}
-
-	if(result > (60 * 24 * 7 * 52))
-		result = (60 * 24 * 7 * 52);
```
Comment (edk0) [inline ircd/s_newconf.c:702]: I think removing this check is out of scope.

File: ircd/s_newconf.c [hunk 3]
```diff
@@ -685,23 +685,39 @@ time_t
...
-	{
-		if(IsDigit(*p))
-		{
-			result *= 10;
-			result += ((*p) & 0xF);
-			p++;
-		}
-		else
+	int current = 0;
+	char *ch = (char *)p, *och;
```
Comment (edk0) [inline ircd/s_newconf.c:689]: I know it'll make the strtol more cumbersome, but I don't think that's a good enough reason to cast.

File: ircd/s_newconf.c [hunk 4]
```diff
@@ -685,23 +685,39 @@ time_t
...
-	while(*p)
-	{
-		if(IsDigit(*p))
-		{
-			result *= 10;
-			result += ((*p) & 0xF);
-			p++;
-		}
-		else
+	int current = 0;
```
Comment (edk0) [inline ircd/s_newconf.c:688]: Wouldn't long be better, since that's what strtol returns? (Or unsigned long and strtoul?)

File: ircd/s_newconf.c [hunk 5]
```diff
@@ -685,23 +685,39 @@ time_t
...
+	char *ch = (char *)p, *och;
+
+	while (*ch) {
+		och = ch;
+		current = strtol(ch, &ch, 10);
```
Comment (edk0) [inline ircd/s_newconf.c:693]: Please check for range errors (return value of LONG_MIN/LONG_MAX + errno == ERANGE)

File: ircd/s_newconf.c [hunk 6]
```diff
@@ -685,23 +685,39 @@ time_t
...
+					break;
+				case 'h':
+					result += (current * 3600);
+					break;
+				case 'd':
+					result += (current * 86400);
+					break;
+				case 'w':
+					result += (current * 604800);
+					break;
```
Comment (edk0) [inline ircd/s_newconf.c:713]: It would make sense to me to check for overflow for these

**Review by edk0 — CHANGES_REQUESTED** [review]
Thanks for the PR! 🍰

File: ircd/s_newconf.c [hunk 7]
```diff
@@ -685,23 +685,53 @@ time_t
...
+				case 'w':
+					if (errno == ERANGE)
+						return -1;
+					result += (current * 604800);
+					break;
+				default:
+					/* Bad time unit was given */
+					return -1;
+			}
```
Comment (dwfreed) [inline ircd/s_newconf.c:729]: All these ERANGE tests are redundant with the one from above.

The null handler will bypass the cap on max duration below.

After fixing the above bypass, this will result in an out of bounds read in the while test if no unit is given for the last number

```suggestion
			switch (*ch) {
				case '\0': /* No unit was given so send it back as minutes */
				case 'm':
					result += (current * 60);
					break;
				case 'h':
					result += (current * 3600);
					break;
				case 'd':
					result += (current * 86400);
					break;
				case 'w':
					result += (current * 604800);
					break;
				default:
					/* Bad time unit was given */
					return -1;
			}
			if (*ch++ == '\0')
				break;
```

File: ircd/s_newconf.c [hunk 8]
```diff
@@ -685,23 +685,53 @@ time_t
...
+		}
 	}
-
 	if(result > (60 * 24 * 7 * 52))
 		result = (60 * 24 * 7 * 52);
```
Comment (dwfreed) [inline ircd/s_newconf.c:733]: Because you're doing the minute unit multiplication before here, you need to include it in the cap test.

```suggestion
	if(result > (60 * 60 * 24 * 7 * 52))
		result = (60 * 60 * 24 * 7 * 52);
```

**Review by dwfreed — CHANGES_REQUESTED** [review]

**Review by jesopo — APPROVED** [review]
wfm, lgtm

**Comment by edk0** [comment]
weird style (`result += (current * 60);`?) aside, this still has a memory leak

**Comment by dwfreed** [comment]
Superseded by #317
